tools/internal/mdocspec/md.go

```diff
@@ -41,53 +37,51 @@ func (r *yamlRenderer) Render(w io.Writer, source []byte, n ast.Node) error {
 
 func (r *yamlRenderer) AddOptions(...renderer.Option) {}
 
-func yamlCodeBlock(source []byte, n ast.Node) *ast.FencedCodeBlock {
-	if s, ok := n.(*ast.FencedCodeBlock); ok && s.Info != nil {
-		if lang := s.Info.Text(source); len(lang) > 0 && string(lang) == "yaml" {
-			return s
+func ocSpecHeading(source []byte, n ast.Node) (heading *ast.Heading, ok bool) {
+	if h, ok := n.(*ast.Heading); ok {
+		headingSegment := h.Lines().At(0)
+		if string(headingSegment.Value(source)) == OCSpecHeading {
+			return h, true
 		}
 	}
-	return nil
+	return nil, false
+}
+
+func yamlCodeBlock(source []byte, n ast.Node) (block *ast.FencedCodeBlock, ok bool) {
+	if c, ok := n.(*ast.FencedCodeBlock); ok && c.Info != nil {
+		if lang := c.Info.Text(source); len(lang) > 0 && string(lang) == "yaml" {
+			return c, true
+		}
+	}
+	return nil, false
 }
 
 func renderYAML(w io.Writer, source []byte, n ast.Node, entering bool) (ast.WalkStatus, error) {
 	if !entering {
 		return ast.WalkContinue, nil
 	}
-	if s := yamlCodeBlock(source, n); s != nil {
-		// Check if within the last `OCSpecHeadingSearchN` siblings,
-		// the "OC Path and RPC Coverage" heading can be found, and
-		// that there is no intervening yaml block.
-		yamlBlockIsOCSpec := false
-		prev := s.PreviousSibling()
-		for i := 0; i != OCSpecHeadingSearchN && prev != nil; i++ {
-			if yamlCodeBlock(source, prev) != nil {
-				break
-			}
-			if h, ok := prev.(*ast.Heading); ok {
-				if h.Lines().Len() > 0 {
-					headingSegment := h.Lines().At(0)
-					if string(headingSegment.Value(source)) == OCSpecHeading {
-						yamlBlockIsOCSpec = true
-						break
-					}
-				}
-			}
-			prev = prev.PreviousSibling()
-		}
-		if !yamlBlockIsOCSpec {
+	heading, ok := ocSpecHeading(source, n)
+	if !ok {
+		return ast.WalkContinue, nil
+	}
+	// Check if prior to the next heading of the same level,
+	// a yaml code block can be found.
+	for next := heading.NextSibling(); next != nil; next = next.NextSibling() {
+		if h, ok := next.(*ast.Heading); ok && h.Level <= heading.Level {
+			// End of heading reached.
 			return ast.WalkContinue, nil
 		}
-
-		l := s.Lines().Len()
-		for i := 0; i != l; i++ {
-			line := s.Lines().At(i)
-			if _, err := w.Write(line.Value(source)); err != nil {
-				return ast.WalkStop, err
+		if c, ok := yamlCodeBlock(source, next); ok {
+			l := c.Lines().Len()
+			for i := 0; i != l; i++ {
+				line := c.Lines().At(i)
+				if _, err := w.Write(line.Value(source)); err != nil {
+					return ast.WalkStop, err
+				}
 			}
+			// Stop after finding the first such YAML block.
+			return ast.WalkStop, nil
```

Comment:
Should we continue parsing and throw an error if there is more than one YAML block?  ie: there must be exactly one YAML block in the ocSpecHeading?   I am not sure we really care if there are other yaml blocks in other headings.

Reply:
I think it's ok to ignore YAML blocks aside from the first one. I don't think it's realistic to achieve the goal of making sure that the author doesn't put multiple coverage-related blocks into the same README. Even if they did, the duplicate block should be caught in review anyways, so I don't think this will be an actual issue. WDYT?

Reply:
Sure, good enough as is